src/query/storages/fuse/src/operations/mutation/compact_segment_mutator.rs

```diff
@@ -0,0 +1,384 @@
+                                counter!("fuse_compact_segments_irreconcilable_conflict", 1);
+                                counter!("fuse_compact_segments_aborts", 1);
+                                abort_segment_compaction(
+                                    &self.data_accessor,
+                                    &self.compacted_segment_accumulator.locations,
+                                )
+                                .await;
+                                break Err(ErrorCode::StorageOther(
+                                    "mutation conflicts, concurrent mutation detected while committing segment compaction operation",
+                                ));
+                            }
+                            Conflict::ReconcilableAppend(r) => {
+                                counter!("fuse_compact_segments_reconcilable_conflict", 1);
+                                info!("resolvable conflicts detected");
+                                concurrently_appended_segments_locations =
+                                    &latest_snapshot.segments[r];
+                            }
+                        }
+
+                        retries += 1;
+
+                        counter!("fuse_compact_segments_retires", retries);
+                        if retries >= MAX_RETRIES {
+                            counter!("fuse_compact_segments_aborts", 1);
+                            abort_segment_compaction(
+                                &self.data_accessor,
+                                &self.compacted_segment_accumulator.locations,
+                            )
+                            .await;
+                            return Err(ErrorCode::StorageOther(format!(
+                                "compact segment failed after {} retries",
+                                retries
+                            )));
+                        }
+                        // different for retry of commit_insert, here operation retired without hesitation
+                        // since... we are in a hurry ;)
+                        continue;
+                    } else {
+                        return Err(e);
+                    }
+                }
+                Ok(_) => return Ok(()),
+            }
+        }
+    }
+}
+
+enum Conflict {
+    Irreconcilable,
+    // reconcilable conflicts with append only operation
+    // the range embedded is the range of segments that are appended in the latest snapshot
+    ReconcilableAppend(Range<usize>),
+}
+
+fn detect_conflicts(base: &TableSnapshot, latest: &TableSnapshot) -> Conflict {
+    let base_segments = &base.segments;
+    let latest_segments = &latest.segments;
+
+    let base_segments_len = base_segments.len();
+    let latest_segments_len = latest_segments.len();
+
+    if latest_segments_len >= base_segments_len
+        && base_segments[base_segments_len - 1] == latest_segments[latest_segments_len - 1]
```

Comment:
Is`base_segments_len` will 0?

Reply:
might be :+1:

Reply:
the logic of conflicts detection of mutators has been unified, in commit.rs, please take another look